### PR TITLE
chore: replaced generators with async/await across the entire codebase

### PR DESCRIPTION
**Pull request by tlakomy**
As mentioned in https://stedi.slab.com/posts/80-jso-nata-evaluation-improvement-notes-s9aqqtce this PR is heavily based on https://github.com/jsonata-js/jsonata/pull/508 since apart from enabling non-blocking/asynchronous functions, it provides significant performance improvements.

Performance benchmark results follow.

Before - **1724ms**
Screenshot 2022-06-07 at 17 03 26

After - **286ms:**
Screenshot 2022-06-07 at 17 17 59

~80% improvement for this particular JSONata expression 🎉 

Closes https://github.com/Stedi/mappings/issues/874

### CODE REVIEW

File: test/run-test-suite.js [hunk 1]
```diff
@@ -180,24 +178,26 @@ function timeboxExpression(expr, timeout, maxDepth) {
...
- * @param {Object} testcase Testcase data read from testcase file
- * @returns {any} The input data to use when evaluating the jsonata expression
- */
+  * Based on the collection of datasets and the information provided as part of the testcase,
+  * determine what input data to use in the case (may return undefined).
+  *
+  * @param {Object} datasets Object mapping dataset names to JS values
+  * @param {Object} testcase Testcase data read from testcase file
+  * @returns {any} The input data to use when evaluating the jsonata expression
+  */
```
Comment (asennikov) [inline test/run-test-suite.js:200]: not sure that is needed

Comment (tlakomy) [reply test/run-test-suite.js:200]: Ah, that would be `prettier`. I'll remove all whitespace changes in this file

File: test/async-function.js [hunk 2]
```diff
@@ -103,83 +93,77 @@ describe('Evaluate concurrent expressions with callbacks', function() {
...
+    const counter = async (count) => ({
+        value: count,
+        inc: async () => await counter(count + 1)
+    });
 
     var bindings = {
         counter: counter
     };
 
     it('basic function that returns a thenable', function() {
```
Comment (asennikov) [inline test/async-function.js:121]: well, I guess it's not "thenable" anymore

File: test/async-function.js [hunk 3]
```diff
@@ -103,83 +93,77 @@ describe('Evaluate concurrent expressions with callbacks', function() {
...
-            },
-            then: function(onFulfilled) {
-                setImmediate(function(){
-                    onFulfilled(count);
-                });
-                return Promise.resolve(count);
-            }
-        };
-    };
+    const counter = async (count) => ({
```
Comment (asennikov) [inline test/async-function.js:112]: curious, why the `counter` function itself is `async`? it looks like only its `inc` property should be `async`?

Comment (asennikov) [reply test/async-function.js:112]: ah, nevermind - it's used within its own property

File: test/async-function.js (hunk 2 above)
Comment (asennikov) [reply test/async-function.js:121]: maybe we could just call it "basic async function" now?

**Comment by RafalWilinski** [comment]
🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢 🚢

File: package.json [hunk 4]
```diff
@@ -13,7 +13,7 @@
   "scripts": {
     "pretest": "npm run lint",
     "mocha": "node ./node_modules/istanbul/lib/cli.js cover --report cobertura --report html ./node_modules/mocha/bin/_mocha -- \"test/**/*.js\"",
-    "test": "npm run mocha",
+    "test": "mocha \"test/run-test-suite.js\"",
```
[the remark this answers is not in the document]
Comment (tlakomy) [reply package.json:16]: Ah, this was a temp change, let me revert it

File: src/utils.js [hunk 5]
```diff
@@ -190,17 +217,19 @@ const utils = (() => {
...
+        isAsyncIterable,
         isDeepEqual,
-        stringToArray
+        isFunction,
+        isIterable,
+        isLambda,
+        isNumeric,
+        isPromise,
+        isSequence,
+        stringToArray,
```
[the remark this answers is not in the document]
Comment (tlakomy) [reply src/utils.js:232]: Good call, let me take that back

**Comment by tlakomy** [comment]
@asennikov @vitorbal @RafalWilinski all comments addressed

File: src/jsonata.js [hunk 6]
```diff
@@ -955,9 +950,9 @@ var jsonata = (function() {
             }
         }
 
-        // iterate over the groups to evaluate the 'value' expression
-        for (key in groups) {
-            entry = groups[key];
+         // iterate over the groups to evaluate the 'value' expression
+         let generators = await Promise.all(Object.keys(groups).map(async (key, idx) => {
```
Comment (asennikov) [inline src/jsonata.js:954]: sorry, could we push the indent back a bit to match the original?

Comment (tlakomy) [reply src/jsonata.js:954]: On it

File: test/implementation-tests.js [hunk 7]
```diff
@@ -12,14 +12,21 @@
...
+    foo: {
+        bar: 42,
+        blah: [
+            { baz: { fud: "hello" } },
+            { baz: { fud: "world" } },
+            { bazz: "gotcha" },
+        ],
+        "blah.baz": "here",
+    },
+    bar: 98,
```
Comment (asennikov) [inline test/implementation-tests.js:29]: I feel like I focus on the wrong thing, but I feel like the prettier changes in this file can make our life harder later on when we try to merge this back

File: test/run-test-suite.js (hunk 1 above)
Comment (asennikov) [inline test/run-test-suite.js:200]: let's move it back where it was (1 less space on the left)

File: test/run-test-suite.js [hunk 8]
```diff
@@ -11,15 +11,17 @@ var path = require("path");
...
- * Simple function to read in JSON
- * @param {string} dir - Directory containing JSON file
- * @param {string} file - Name of JSON file (relative to directory)
- * @returns {Object} Parsed JSON object
- */
+  * Simple function to read in JSON
+  * @param {string} dir - Directory containing JSON file
+  * @param {string} file - Name of JSON file (relative to directory)
+  * @returns {Object} Parsed JSON object
+  */
```
Comment (asennikov) [inline test/run-test-suite.js:24]: another case we could move it back

File: test/parser-recovery.js [hunk 9]
```diff
@@ -308,87 +308,87 @@ describe('Invoke parser with incomplete expression', function() {
...
-                            "position": 36
-                        }
-                    ]
-                };
+                            }
+                        ],
+                        "position": 36
+                    }
+                ]
+            };
```
Comment (asennikov) [inline test/parser-recovery.js:391]: should we push this object 4 spaces to the right to match the original?

Comment (tlakomy) [reply test/parser-recovery.js:391]: Trying to make that happen, not sure why my editor reverts those changes

**Review by asennikov — APPROVED** [review]
merge away 🚢

**Pull request merged**